docs/CppInteroperability/SwiftToCppInteroperabilityRoadmap.md

```diff
@@ -0,0 +1,125 @@
+
+The existing model of generating a header file that can then be imported by Objective-C code can be adopted to bridge Swift declarations to C++ as well. The Clang compiler knows how to emit calls to functions with the Swift calling convention, so it’s possible to call into Swift from C++ using the matching ABI that the Swift program expects. This in turn allows the Swift compiler to generate a C++ header that is able to call into Swift functions, methods, initializers and accessors correctly. These raw calls that use the Swift calling convention can then be wrapped in inline thunk stub functions, allowing a C++ programmer to call into Swift APIs using the familiar C++ function and member function call syntax. Swift types can also be mapped to C++ types, allowing a C++ programmer to use and manipulate Swift values using familiar syntax from C++. The exact details of how Swift types and other language constructs get mapped to C++ are not covered in this document, as there’s another document that describes this: https://github.com/apple/swift/blob/main/docs/CppInteroperability/UserGuide-CallingSwiftFromC%2B%2B.md .
+
+## Goals
+
+Swift and C++ are both feature rich programming languages. While they both have a lot of similar features, Swift has some features which C++ can not represent natively. For instance, Swift supports overloading based on argument labels, but C++ does not. A gap like this can be bridged in an idiomatic manner between the two languages, as the C++ name of the exposed Swift function can be modified to reflect the Swift argument labels. However, some Swift features like result builders can not be bridged in an idiomatic manner to C++. Therefore, we can assume that it is impractical for Swift to expose all Swift APIs to C++. Additionally, some Swift API patterns, like conforming Swift types to specific protocols, require advanced compiler features that are impractical to support in a generated header. Thus, we can assume that it is also impractical for Swift to bridge all Swift API patterns into C++.
+
+For the reasons outlined above, it is not a goal to expose every Swift API and language construct to C++, and it is also not a goal to make every Swift library accessible to C++. There will certainly be libraries that will lend themselves well to an ergonomic and an idiomatic C++ interface, however, there will always be some libraries that rely on language features or API patterns that can’t be bridged to C++ in a reasonable manner. 
```

Comment:
Ideally users would not be encouraged to write Swift APIs using a subset of the language that mapped well to C++ in hopes of vending a nice C++ API interface. Mostly, this is not something we can prevent. But maybe there's a way to call that out here, or keep it in mind. I think it's a hard goal to express, though, so maybe best to just leave it out.